fodselsnummer.py

```diff
@@ -1,54 +1,94 @@
-
+    Check if a number is a valid fødselsnumber.
     Args:
         fnr: A string containing the fodselsnummer to check
+        h_numbers: False (the default) if h-numbers should be accepted
         d_numbers: True (the default) if d-numbers should be accepted
     Returns:
         True if it is a valid fodselsnummer, False otherwise.
     """
     if not FNR_REGEX.match(fnr):
-        return False
+        raise ValueError('Fødselsnumber does not match regex \\d{11}')
+
+    individual_number = int(fnr[6:9])
+    day, month, year = int(fnr[0:2]), int(fnr[2:4]), int(fnr[4:6])
+    if 41 <= day <= 71:  # if D-number
+        if not d_numbers:
+            raise ValueError('Fødselsnumber is a D-number (or day out of range)')
+        day -= 40
 
+    if not (1 <= day <= 31):
+        raise ValueError('Day out of range in fødselsnumber')
+    if 41 <= month <= 52:  # if H-number
+        if h_numbers:
+            month -= 40
+        else:
+            raise ValueError('Fødselsnumber is a H-number (or month out of range)')
+    if not 1 <= month <= 12:
+        raise ValueError('Month out of range in fødselsnumber')
+    if individual_number <= 499:  # individual numbers 000-499 indicate person is born in 19XX
+        year += 1900
+    else:
+        year += 2000
+    if individual_number >= 900 and year >= 2040:
+        year -= 100
+    dob: Optional[date] = None
     try:
-        datetime.strptime(fnr[0:6], '%d%m%y')
+        dob = date(year=year, month=month, day=day)
+        if dob > date.today():
+            raise ValueError('Date of fødselsnumber is larger than current date')
     except ValueError:
-        if d_numbers:
-            dnrdatestring = str(int(fnr[0])-4) + fnr[1:6]
-            try:
-                datetime.strptime(dnrdatestring, '%d%m%y')
-            except ValueError:
-                return False
-        else:
-            return False
+        raise ValueError(f'{year}-{month}-{day} is not a valid date')
 
     try:
         generatedfnr = _generate_control_digits(fnr[0:9])
     except InvalidControlDigitException:
+        raise ValueError('Control digit does not match fødselsnumber')
+
+    if not bool(fnr == generatedfnr):
+        raise ValueError('Control digit does not match fødselsnumber')
+    return True
+
+
+def check_fnr(fnr: str, d_numbers=True, h_numbers=False, logger: Callable = lambda _x: None) -> bool:
+    """
+    Check if a number is a valid fødselsnumber.
+    Args:
+        fnr: A string containing the fodselsnummer to check
+        h_numbers: False (the default) if h-numbers should be accepted
+        d_numbers: True (the default) if d-numbers should be accepted
+        logger: A function used to log things
+    Returns:
+        True if it is a valid fodselsnummer, False otherwise.
+    """
+    try:
+        return validate_fnr(fnr=fnr, d_numbers=d_numbers, h_numbers=h_numbers)
+    except Exception as e:
+        logger(next((a for a in e.args), ''))
```

Comment:
Why not just?
```
logger(str(e))
```

Reply:
with `str(e)` you will get nice error text for `OSError` (e.g. "[Errno 2] No such file or directory: 'foo'"), but with current solution you will get first item from exception args which is just int value for errno (e.g. "2").